zstash/globus_utils.py

```diff
@@ -0,0 +1,273 @@
+            for ep in endpoints
+            if ep is not None
+        ]
+    )
+    return f"urn:globus:auth:scope:transfer.api.globus.org:all[{inner}]"
+
+
+def save_tokens(token_response):
+    tokens_to_save = {}
+    for resource_server, token_data in token_response.by_resource_server.items():
+        tokens_to_save[resource_server] = {
+            "access_token": token_data["access_token"],
+            "refresh_token": token_data.get("refresh_token"),
+            "expires_at": token_data.get("expires_at_seconds"),
+        }
+
+    with open(TOKEN_FILE, "w") as f:
+        json.dump(tokens_to_save, f, indent=2)
+    logger.info("Tokens saved successfully")
+
+
+# Primarily used by globus_transfer ###########################################
+def set_up_TransferData(
+    transfer_type: str,
+    local_endpoint: Optional[str],
+    remote_endpoint: Optional[str],
+    remote_path: str,
+    name: str,
+    transfer_client: TransferClient,
+    transfer_data: Optional[TransferData] = None,
+) -> TransferData:
+    if not local_endpoint:
+        raise ValueError("Local endpoint ID is not set.")
+    if not remote_endpoint:
+        raise ValueError("Remote endpoint ID is not set.")
+    if transfer_type == "get":
+        src_ep = remote_endpoint
+        src_path = os.path.join(remote_path, name)
+        dst_ep = local_endpoint
+        dst_path = os.path.join(os.getcwd(), name)
+    else:
+        src_ep = local_endpoint
+        src_path = os.path.join(os.getcwd(), name)
+        dst_ep = remote_endpoint
+        dst_path = os.path.join(remote_path, name)
+
+    subdir = os.path.basename(os.path.normpath(remote_path))
+    subdir_label = re.sub("[^A-Za-z0-9_ -]", "", subdir)
+    filename = name.split(".")[0]
+    label = subdir_label + " " + filename
+
+    if not transfer_data:
+        transfer_data = TransferData(
+            transfer_client,
+            src_ep,
+            dst_ep,
+            label=label,
+            verify_checksum=True,
+            preserve_timestamp=True,
+            fail_on_quota_errors=True,
+        )
+    transfer_data.add_item(src_path, dst_path)
+    transfer_data["label"] = label
+    return transfer_data
+
+
+def submit_transfer_with_checks(transfer_client, transfer_data) -> GlobusHTTPResponse:
+    task: GlobusHTTPResponse
+    try:
+        task = transfer_client.submit_transfer(transfer_data)
+    except TransferAPIError as err:
+        if err.info.consent_required:
+            logger.error("Consent required - this suggests scope issues.")
+            logger.error(
+                "With proper scope handling, this block should not be reached."
+            )
+            logger.error(
+                "Please report this bug at https://github.com/E3SM-Project/zstash/issues, with details of what you were trying to do."
+            )
+            raise RuntimeError(
+                "Insufficient Globus consents - please report this bug"
+            ) from err
```

Comment:
I left a check in this block, but it's a little unclear if/how it would be reached now, so I figured it would be best to just leave a debugging note here.